lib/realtime/server.ex

```diff
@@ -149,32 +171,67 @@ defmodule Realtime.Server do
       ) do
     new_active_route_ids = Map.keys(vehicles_by_route_id)
 
-    _ = update_vehicle_positions(state, vehicles_by_route_id, shuttles, new_active_route_ids)
+    new_active_trip_ids =
+      vehicles_by_route_id
+      |> Map.values()
+      |> List.flatten()
+      |> Enum.map(& &1.trip_id)
+      |> Enum.filter(& &1)
+
+    _ =
+      update_vehicle_positions(
+        state,
+        vehicles_by_route_id,
+        shuttles,
+        new_active_route_ids,
+        new_active_trip_ids
+      )
+
+    new_state =
+      Map.merge(state, %{
+        active_route_ids: new_active_route_ids,
+        active_trip_ids: new_active_trip_ids
+      })
 
-    new_state = Map.put(state, :active_route_ids, new_active_route_ids)
     _ = broadcast(new_state)
 
     {:noreply, new_state}
   end
 
   defp update_vehicle_positions(
-         %__MODULE__{ets: ets, active_route_ids: active_route_ids},
+         %__MODULE__{
+           ets: ets,
+           active_route_ids: active_route_ids,
+           active_trip_ids: active_trip_ids
+         },
          vehicles_by_route_id,
          shuttles,
-         new_active_route_ids
+         new_active_route_ids,
+         new_active_trip_ids
        ) do
     removed_route_ids = active_route_ids -- new_active_route_ids
+    removed_trip_ids = active_trip_ids -- new_active_trip_ids
 
     for route_id <- removed_route_ids do
       _ = :ets.delete(ets, {:route_id, route_id})
     end
 
+    for trip_id <- removed_trip_ids do
+      _ = :ets.delete(ets, {:trip_id, trip_id})
+    end
+
     for {route_id, vehicles_and_ghosts} <- vehicles_by_route_id do
       active_vehicles_and_ghosts = Enum.filter(vehicles_and_ghosts, &block_is_active?/1)
       _ = :ets.insert(ets, {{:route_id, route_id}, active_vehicles_and_ghosts})
     end
 
-    :ets.insert(ets, {:all_vehicles, Enum.uniq(all_vehicles(vehicles_by_route_id) ++ shuttles)})
+    all_vehicles = Enum.uniq(all_vehicles(vehicles_by_route_id) ++ shuttles)
+
+    for vehicle <- all_vehicles do
+      _ = :ets.insert(ets, {{:trip_id, vehicle.trip_id}, vehicle})
```

Comment:
What happens if two vehicles have the same trip id? (It shouldn't happen, since overloads get an `ADDED` trip id, but I just want to make sure nothing will explode.)

Reply:
`ets` has type `set`, so the last `vehicle` iterated over in that case is the one recorded. But as you say, shouldn't happen.